icepll/icepll.cc

```diff
@@ -157,29 +182,114 @@ int main(int argc, char **argv)
+		printf("DIVF: %2d (7'b%s)\n", best_divf, binstr(best_divf, 7));
+		printf("DIVQ: %2d (3'b%s)\n", best_divq, binstr(best_divq, 3));
 
-	printf("\n");
+		printf("\n");
 
-	printf("FILTER_RANGE: %d (3'b%s)\n", filter_range, binstr(filter_range, 3));
+		printf("FILTER_RANGE: %d (3'b%s)\n", filter_range, binstr(filter_range, 3));
 
-	printf("\n");
+		printf("\n");
+	}
+
+	// save PLL configuration as file
+	if (filename != NULL)
+	{
+		// open file for writing
+		FILE *f;
+		f = fopen(filename, "w");
+
+		if (save_as_module)
+		{
+			// save PLL configuration as Verilog module
+
+			// header
+			fprintf(f, "/**\n * PLL configuration\n *\n"
+						" * This Verilog module was generated automatically\n"
+						" * using the icepll tool from the IceStorm project.\n"
+						" * Use at your own risk.\n"
+						" *\n"
+						" * Given input frequency:      %8.3f MHz\n"
+						" * Requested output frequency: %8.3f MHz\n"
+						" * Achieved output frequency:  %8.3f MHz\n"
+						" */\n\n",
+						f_pllin, f_pllout, best_fout);
+
+			// generate Verilog module
+			fprintf(f,  "module pll(\n"
+						"\tinput  clock_in,\n"
+						"\toutput clock_out,\n"
+						"\toutput locked\n"
+						"\t);\n\n"
+					);
+
+			// save iCE40 PLL tile configuration
+			fprintf(f, "SB_PLL40_CORE #(\n");
+			fprintf(f, "\t\t.FEEDBACK_PATH(\"%s\"),\n", (simple_feedback ? "SIMPLE" : "NON_SIMPLE"));
+			fprintf(f, "\t\t.PLLOUT_SELECT(\"GENCLK\"),\n");
+			fprintf(f, "\t\t.DIVR(4'b%s),\t\t"      "// DIVR = %2d\n", binstr(best_divr, 4), best_divr);
+			fprintf(f, "\t\t.DIVF(7'b%s),\t"        "// DIVF = %2d\n", binstr(best_divf, 7), best_divf);
+			fprintf(f, "\t\t.DIVQ(3'b%s),\t\t"      "// DIVQ = %2d\n", binstr(best_divq, 3), best_divq);
+			fprintf(f, "\t\t.FILTER_RANGE(3'b%s)\t" "// FILTER_RANGE = %d\n", binstr(filter_range, 3), filter_range);
+			fprintf(f, "\t) uut (\n"
+						"\t\t.LOCK(locked),\n"
+						"\t\t.RESETB(1'b1),\n"
+						"\t\t.BYPASS(1'b0),\n"
+						"\t\t.REFERENCECLK(clock_in),\n"
+						"\t\t.PLLOUTCORE(clock_out),\n"
+						"\t\t);\n\n"
+					);
+
+			fprintf(f, "endmodule\n");
+		}
+		else
+		{
+			// only save PLL configuration values
+
+			// header
+			fprintf(f, "/**\n * PLL configuration\n *\n"
+						" * This Verilog header file was generated automatically\n"
+						" * using the icepll tool from the IceStorm project.\n"
+						" * It is intended for use with FPGA primitives SB_PLL40_CORE,\n"
+						" * SB_PLL40_PAD, SB_PLL40_2_PAD, SB_PLL40_2F_CORE or SB_PLL40_2F_PAD.\n"
+						" * Use at your own risk.\n"
+						" *\n"
+						" * Given input frequency:      %8.3f MHz\n"
+						" * Requested output frequency: %8.3f MHz\n"
+						" * Achieved output frequency:  %8.3f MHz\n"
+						" */\n\n",
+						f_pllin, f_pllout, best_fout);
+
+			// PLL configuration
+			fprintf(f, ".FEEDBACK_PATH(\"%s\"),\n", (simple_feedback ? "SIMPLE" : "NON_SIMPLE"));
+			fprintf(f, ".PLLOUT_SELECT(\"GENCLK\"),\n");
```

Comment:
`PLLOUT_SELECT` is specific to the `SB_PLL40_CORE` cell (some of the other PLL cell types have `PLLOUT_SELECT_PORTA` and `PLLOUT_SELECT_PORTB`). Also: there is no reason to limit to user to the `GENCLK` option, and `GENCLK` is the default anyways. So I think this like should be removed.